File: src/app/VoiceBooth.cpp
#include "VoiceBooth.h"

#include "core/LogModel.h"
#include "core/Paths.h"
#include "core/Pricing.h"
#include "core/SettingsStore.h"
#include "providers/ProviderTask.h"
#include "providers/Registry.h"
#include "providers/Types.h"
#include "providers/VoiceProviders.h"

#include <QDateTime>
#include <QDir>
#include <QFile>

namespace {

// ElevenLabs bills text-to-speech per thousand characters.
constexpr double kCharsPerKilo = 1000.0;

double settingOr(const QVariantMap &actor, const char *key, double fallback)
{
    const QVariant value = actor.value(QString::fromLatin1(key));
    return value.isValid() ? value.toDouble() : fallback;
}

} // namespace

VoiceBooth::VoiceBooth(SettingsStore *settings, Registry *registry, Pricing *pricing, LogModel *log,
                       QObject *parent)
    : QObject(parent)
    , m_settings(settings)
    , m_registry(registry)
    , m_pricing(pricing)
    , m_log(log)
{
}

QString VoiceBooth::scratchDir() const
{
    return paths::ensureDir(QDir(paths::configDir()).filePath(QStringLiteral("auditions")));
}

void VoiceBooth::setError(const QString &error)
{
    if (m_error == error)
        return;
    m_error = error;
    emit errorChanged();
}

void VoiceBooth::setAuditioning(bool on)
{
    if (m_auditioning == on)
        return;
    m_auditioning = on;
    emit auditioningChanged();
}

void VoiceBooth::setCloning(bool on)
{
    if (m_cloning == on)
        return;
    m_cloning = on;
    emit cloningChanged();
}

void VoiceBooth::cancel()
{
    if (m_task && !m_task->isFinished())
        m_task->cancel();
    m_task = nullptr;
    setAuditioning(false);
    setCloning(false);
}

// ---------------------------------------------------------------------------
// Audition
// ---------------------------------------------------------------------------
QVariantMap VoiceBooth::estimate(const QString &text) const
{
    QString providerId = m_settings->pref(QStringLiteral("voiceProvider")).toString();
    if (providerId.isEmpty())
        providerId = m_registry->defaultProvider(QStringLiteral("voice"));

    const QString model = m_registry->resolveModel(
        providerId, m_settings->pref(QStringLiteral("voiceModel")).toString());

    const QVariantMap unit = m_pricing->unitPrice(model);
    if (unit.isEmpty())
        return {{QStringLiteral("known"), false}, {QStringLiteral("amount"), 0.0}};

    const double kilos = text.size() / kCharsPerKilo;
    return {{QStringLiteral("known"), true},
            {QStringLiteral("amount"), unit.value(QStringLiteral("amount")).toDouble() * kilos}};
}

void VoiceBooth::audition(const QVariantMap &actor, const QString &text)
{
    if (m_auditioning || m_cloning)
        return;

    const QString line = text.trimmed();
    if (line.isEmpty()) {
        setError(tr("Write a line for them to say."));
        return;
    }

    const QString dir = scratchDir();
    if (dir.isEmpty()) {
        setError(tr("Could not create the auditions folder."));
        return;
    }

    QString providerId = m_settings->pref(QStringLiteral("voiceProvider")).toString();
    if (providerId.isEmpty())
        providerId = m_registry->defaultProvider(QStringLiteral("voice"));

    prov::VoiceRequest request;
    request.apiKey = m_settings->apiKey(m_registry->credentialFor(providerId));
    request.model = m_registry->resolveModel(
        providerId, m_settings->pref(QStringLiteral("voiceModel")).toString());
    request.voiceId = actor.value(QStringLiteral("voiceId")).toString();
    request.text = line;
    request.stability = settingOr(actor, "voiceStability", 0.45);
    request.similarity = settingOr(actor, "voiceSimilarity", 0.8);
    request.style = settingOr(actor, "voiceStyle", 0.35);
    request.speed = settingOr(actor, "voiceSpeed", 1.0);

    ProviderTask *task = providers::voice(providerId, request, this);
    if (!task) {
        setError(tr("No voice provider called %1.").arg(providerId));
        return;
    }

    setError({});
    setAuditioning(true);
    m_task = task;

    connect(task, &ProviderTask::failed, this, [this](const QString &error) {
        m_log->append(LogModel::Error, tr("Audition failed: %1").arg(error));
        setError(error);
        setAuditioning(false);
    });

    connect(task, &ProviderTask::succeeded, this, [this, dir](const QVariantMap &result) {
        const QByteArray data = result.value(QStringLiteral("data")).toByteArray();
        const QString extension =
            result.value(QStringLiteral("extension"), QStringLiteral("mp3")).toString();

        // A fresh name every time: an audio player that already has the file
        // open will not reload one that kept its path.
        const QString path =
            QDir(dir).filePath(QStringLiteral("audition-%1.%2")
                                   .arg(QDateTime::currentDateTime().toString(
                                            QStringLiteral("yyyyMMdd-HHmmss-zzz")),
                                        extension));

        QFile file(path);
        if (!file.open(QIODevice::WriteOnly | QIODevice::Truncate)) {
            setError(tr("Could not write %1.").arg(QDir::toNativeSeparators(path)));
            setAuditioning(false);
            return;
        }
        file.write(data);
        file.close();

        m_samplePath = path;
        emit samplePathChanged();
        setAuditioning(false);
    });
}

// ---------------------------------------------------------------------------
// Cloning
// ---------------------------------------------------------------------------
void VoiceBooth::addCloneSamples(const QList<QUrl> &urls)
{
    const int before = m_cloneSamples.size();
    for (const QUrl &url : urls) {
        const QString path = url.isLocalFile() ? url.toLocalFile() : url.toString();
        if (!path.isEmpty() && !m_cloneSamples.contains(path))
            m_cloneSamples.append(path);
    }
    if (m_cloneSamples.size() != before)
        emit cloneSamplesChanged();
}

void VoiceBooth::removeCloneSample(int index)
{
    if (index < 0 || index >= m_cloneSamples.size())
        return;
    m_cloneSamples.removeAt(index);
    emit cloneSamplesChanged();
}

void VoiceBooth::cloneVoice(const QString &name)
{
    if (m_auditioning || m_cloning)
        return;

    if (m_cloneSamples.isEmpty()) {
        setError(tr("Add at least one recording first."));
        return;
    }

    prov::VoiceCloneRequest request;
    request.apiKey = m_settings->apiKey(m_registry->credentialFor(QStringLiteral("elevenlabs")));
    request.name = name.trimmed().isEmpty() ? tr("Market Queen voice") : name.trimmed();
    request.description = tr("Cloned with Market Queen.");
    request.samplePaths = m_cloneSamples;

    auto *task = new ElevenLabsVoiceCloneTask(request, this);

    setError({});
    setCloning(true);
    m_task = task;

    m_log->append(LogModel::Info,
                  tr("Cloning a voice from %1 recording(s).").arg(m_cloneSamples.size()));

    connect(task, &ProviderTask::failed, this, [this](const QString &error) {
        m_log->append(LogModel::Error, tr("Cloning failed: %1").arg(error));
        setError(error);
        setCloning(false);
    });

    connect(task, &ProviderTask::succeeded, this, [this](const QVariantMap &result) {
        const QString voiceId = result.value(QStringLiteral("voiceId")).toString();
        const QString name = result.value(QStringLiteral("name")).toString();
        m_log->append(LogModel::Success, tr("Voice \"%1\" is on your account.").arg(name));
        setCloning(false);
        emit cloned(voiceId, name);
    });
}

File: market_queen_flutter_app/lib/models/voice_booth.dart
import 'dart:io';

import 'package:flutter/foundation.dart';
import 'package:path/path.dart' as p;

import '../core/http_util.dart';
import '../core/log_model.dart';
import '../core/paths.dart';
import '../core/pricing.dart';
import '../core/settings_store.dart';
import '../core/signal.dart';
import '../i18n/translator.dart';
import '../providers/provider_task.dart';
import '../providers/registry.dart';
import '../providers/types.dart';
import '../providers/voice_casting.dart';
import '../providers/voice_providers.dart';

/// The audition booth.
///
/// A voice-over is the cheapest part of an ad to get wrong and the most obvious
/// once it is: an announcer read sinks a UGC ad as surely as a model's face. So
/// this exists to make hearing the actor cost a fraction of a cent, before any
/// video is bought.
///
/// Cloning is deliberately separate and never implicit. It uploads the user's
/// recordings and creates a permanent voice on their provider account, so it
/// only ever happens on an explicit click.
class VoiceBooth extends ChangeNotifier {
  VoiceBooth(
    this._settings,
    this._registry,
    this._pricing,
    this._log,
    this._casting,
  );

  /// ElevenLabs bills text-to-speech per thousand characters.
  static const _charsPerKilo = 1000.0;

  final SettingsStore _settings;
  final Registry _registry;
  final Pricing _pricing;
  final LogModel _log;
  final VoiceCasting _casting;

  /// The new voice is on the account; the caller decides whether to adopt it.
  final Event<({String voiceId, String name})> cloned = Event();

  ProviderTask? _task;

  String _samplePath = '';
  String _error = '';
  final List<String> _cloneSamples = [];
  bool _auditioning = false;
  bool _cloning = false;

  bool get auditioning => _auditioning;
  bool get cloning => _cloning;

  /// Local file of the last audition, empty until one lands.
  String get samplePath => _samplePath;
  String get error => _error;

  /// Recordings to clone from. Kept here rather than in the project: they are
  /// an input to one operation, not part of the ad.
  List<String> get cloneSamples => List.unmodifiable(_cloneSamples);

  String get _scratchDir => Paths.ensureDir(p.join(Paths.configDir, 'auditions'));

  void _setError(String value) {
    if (_error == value) return;
    _error = value;
    notifyListeners();
  }

  void cancel() {
    final task = _task;
    if (task != null && !task.isFinished) task.cancel();
    _task = null;
    _auditioning = false;
    _cloning = false;
    notifyListeners();
  }

  // ---- Audition ---------------------------------------------------------

  /// What one audition of [text] costs.
  CostEstimate estimate(String text) {
    var providerId = _settings.prefString('voiceProvider');
    if (providerId.isEmpty) providerId = _registry.defaultProvider('voice');

    final model =
        _registry.resolveModel(providerId, _settings.prefString('voiceModel'));

    final unit = _pricing.unitPrice(model);
    if (unit == null) return CostEstimate.unknown;
    return CostEstimate(true, unit.amount * (text.length / _charsPerKilo));
  }

  /// Speaks [text] as the actor would. [actor] is AdProject's actor map: the
  /// casting brief -- or a pinned voice id -- plus the four settings.
  ///
  /// The voice is worked out here rather than passed in, because the whole
  /// point of the audition is that it buys what the render will: the brief goes
  /// through the same caster, hits the same remembered answer, and the actor you
  /// hear is the actor the ad gets.
  Future<void> audition(Map<String, Object?> actor, String text) async {
    if (_auditioning || _cloning) return;

    final line = text.trim();
    if (line.isEmpty) {
      _setError(tr('Write a line for them to say.'));
      return;
    }

    final dir = _scratchDir;
    if (dir.isEmpty) {
      _setError(tr('Could not create the auditions folder.'));
      return;
    }

    var providerId = _settings.prefString('voiceProvider');
    if (providerId.isEmpty) providerId = _registry.defaultProvider('voice');

    final apiKey = _settings.apiKey(_registry.credentialFor(providerId));
    final model =
        _registry.resolveModel(providerId, _settings.prefString('voiceModel'));

    double setting(String key, double fallback) {
      final value = actor[key];
      return value is num ? value.toDouble() : fallback;
    }

    _error = '';
    _auditioning = true;
    notifyListeners();

    try {
      final voice = await _casting.voiceFor(
        providerId: providerId,
        apiKey: apiKey,
        modelId: model,
        source: actor,
        onTask: (running) => _task = running,
      );
      if (voice.label.isNotEmpty) {
        _log.info(voice.description.isEmpty
            //: %1 is a voice's name
            ? tr('Voice: %1.').arg(voice.label)
            //: %1 is a voice's name, %2 what it sounds like
            : tr('Voice: %1 — %2.').arg(voice.label).arg(voice.description));
      }

      final task = ProviderFactory.voice(
        providerId,
        VoiceRequest(
          apiKey: apiKey,
          model: model,
          voiceId: voice.id,
          text: line,
          stability: setting('voiceStability', 0.45),
          similarity: setting('voiceSimilarity', 0.8),
          style: setting('voiceStyle', 0.35),
          speed: setting('voiceSpeed', 1.0),
        ),
      );

      if (task == null) {
        _setError(tr('No voice provider called %1.').arg(providerId));
        _auditioning = false;
        notifyListeners();
        return;
      }
      _task = task;

      final result = await task.run();

      final data = result['data'] as Uint8List? ?? Uint8List(0);
      final extension = '${result['extension'] ?? 'mp3'}';

      // A fresh name every time: an audio player that already has the file open
      // will not reload one that kept its path.
      final path = p.join(dir, 'audition-${_stamp(DateTime.now())}.$extension');

      try {
        File(path).writeAsBytesSync(data);
      } on FileSystemException {
        _setError(tr('Could not write %1.').arg(path));
        _auditioning = false;
        notifyListeners();
        return;
      }

      _samplePath = path;
      _auditioning = false;
      notifyListeners();
    } on ProviderException catch (error) {
      if (error is! TaskCancelled) {
        _log.error(tr('Audition failed: %1').arg(error.message));
        _error = error.message;
      }
      _auditioning = false;
      notifyListeners();
    } finally {
      _task = null;
    }
  }

  // ---- Cloning ----------------------------------------------------------

  void addCloneSamples(List<String> paths) {
    final before = _cloneSamples.length;
    for (final raw in paths) {
      final path = Http.toLocalPath(raw);
      if (path.isNotEmpty && !_cloneSamples.contains(path)) _cloneSamples.add(path);
    }
    if (_cloneSamples.length != before) notifyListeners();
  }

  void removeCloneSample(int index) {
    if (index < 0 || index >= _cloneSamples.length) return;
    _cloneSamples.removeAt(index);
    notifyListeners();
  }

  /// Uploads the samples and creates a voice on the user's account.
  Future<void> cloneVoice(String name) async {
    if (_auditioning || _cloning) return;

    if (_cloneSamples.isEmpty) {
      _setError(tr('Add at least one recording first.'));
      return;
    }

    final task = ElevenLabsVoiceCloneTask(VoiceCloneRequest(
      apiKey: _settings.apiKey(_registry.credentialFor('elevenlabs')),
      name: name.trim().isEmpty ? tr('Market Queen voice') : name.trim(),
      description: tr('Cloned with Market Queen.'),
      samplePaths: List.of(_cloneSamples),
    ));

    _error = '';
    _cloning = true;
    _task = task;
    notifyListeners();

    _log.info(
        tr('Cloning a voice from %1 recording(s).').arg(_cloneSamples.length));

    try {
      final result = await task.run();
      final voiceId = '${result['voiceId'] ?? ''}';
      final voiceName = '${result['name'] ?? ''}';

      _log.success(tr('Voice "%1" is on your account.').arg(voiceName));
      _cloning = false;
      notifyListeners();
      cloned.emit((voiceId: voiceId, name: voiceName));
    } on ProviderException catch (error) {
      if (error is! TaskCancelled) {
        _log.error(tr('Cloning failed: %1').arg(error.message));
        _error = error.message;
      }
      _cloning = false;
      notifyListeners();
    } finally {
      _task = null;
    }
  }

  static String _stamp(DateTime time) {
    String two(int value) => value.toString().padLeft(2, '0');
    String three(int value) => value.toString().padLeft(3, '0');
    return '${time.year}${two(time.month)}${two(time.day)}'
        '-${two(time.hour)}${two(time.minute)}${two(time.second)}'
        '-${three(time.millisecond)}';
  }
}
